### ISSUE TITLE
Discrepancy between `prediction$score` (which returns the implausible C-index of` 1`) and the contents in `learner$model`

### ISSUE THREAD

**Issue by a-humble-learner** [mlr-org/mlr3proba#234]
There is a discrepancy between `prediction$score` (which returns the implausible C-index of` 1`) and the contents in `learner$model`. This problem happens not just when using `glmnet` but also when using `xgboost`. 

A less important point here is `learner$importance()` doesn't seem to work here.

Please see below the reprex and find the fake data that reproduces this error attached.

Thank you.

``` r
ls_library <- list("reprex", "mlr3", "mlr3proba", "mlr3learners","glmnet", "arrow", "caret", "corrplot", "data.table", "dplyr", "forestmodel", "ggplot2", "ggpubr", "gtools", "lubridate", "purrr", "readstata13", "readxl", "reshape2", "scales", "skimr", "stringi", "survival", "tidyverse", "zoo")
lapply(ls_library, function(lib) {
    library(lib, character.only = TRUE)
})

# mk data matrix ----
df <- fread("~/silly_data.csv")
df$cov1_ <- as.factor(df$cov1_)
mat <- model.matrix(~., df, na.action=na.omit)[,-1] %>% as.data.frame() #one-hot-encode factors 
mat %>% str()
#> 'data.frame':    24800 obs. of  6 variables:
#>  $ futime: num  0.347 1.839 0.851 0.306 0.687 ...
#>  $ status: num  1 1 1 1 1 1 1 1 1 1 ...
#>  $ cov1_2: num  0 0 0 0 0 0 0 0 0 0 ...
#>  $ cov1_3: num  0 0 0 0 0 0 0 0 0 0 ...
#>  $ cov1_4: num  0 0 0 0 0 0 0 0 0 0 ...
#>  $ cov2_ : num  1 0 0 1 0 0 0 0 1 1 ...

task = TaskSurv$new(
    id="testing",
    backend=as.data.frame(mat),
    time = "futime",
    event = "status"
    # type = "interval"
    )

train_rownums <- 1:task$nrow


# modelling ----
learner = lrn("surv.cv_glmnet", alpha=1, nfolds=3, type.measure = "C") #or mlr_learners$get("surv.xgboost")

# train ----
learner$train(task, row_ids = train_rownums) 
cat("\n\nlearner$model is ......\n")
#> 
#> 
#> learner$model is ......
learner$model %>% print()
#> 
#> Call:  (if (cv) glmnet::cv.glmnet else glmnet::glmnet)(x = data, y = target,      type.measure = "C", nfolds = 3L, alpha = 1, family = "cox") 
#> 
#> Measure: C-index 
#> 
#>       Lambda Index Measure       SE Nonzero
#> min 0.005266    21  0.7222 0.001817       4
#> 1se 0.005779    20  0.7217 0.001795       4
learner$model %>% coef() %>% print()
#> 4 x 1 sparse Matrix of class "dgCMatrix"
#>                1
#> cov1_2 0.3735680
#> cov1_3 1.2242297
#> cov1_4 1.4016563
#> cov2_  0.2635423

# feature importance ----
learner$importance() #not applicable to glmnet
#> Error in eval(expr, envir, enclos): attempt to apply non-function


#predict ----
prediction = learner$predict(task, row_ids = train_rownums)
cat("\n\nprediction$score() is ......\n")
#> 
#> 
#> prediction$score() is ......
prediction$score() %>% print()
#> surv.harrell_c 
#>              1
```

<sup>Created on 2021-12-20 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>
[silly_data.csv](https://github.com/mlr-org/mlr3proba/files/7746674/silly_data.csv)

**Comment by RaphaelS1** [comment]
Hey, okay a few points (mostly guesses for now)

1. There isn't a discrepancy between `$model` and `$score`. The first is showing you the c-index under different lambda conditions the second is showing you the c-index calculated by comparing the predicted risk and the true times
2. We should implement `$importance` thanks for pointing that out
3. I'm not sure if `C = 1` actually is implausible. First of all you are not splitting training and testing data so your model knows the "exact" relationship between the covariates and targets on a very large dataset. Secondly the c-index works by comparing pairs of observations but critically it removes all pairs where the smaller event time is censored, which is most in your case. Therefore the model only needs to make a handful of predictions with correct discimination to achieve a c-index of 1

**Comment by a-humble-learner** [comment]
Thanks for your comment, @RaphaelS1 .

1. Aren't the C-indices for the different values of lambda  within `$model `also the predictions are done by comparing the predicted risk and the true times? When `glmnet` does its prediction, it uses by default  ` lambda 1se`. In this case, that would simply be lambda= 0.005779. So I am expecting them to agree. Is there something I am missing here?
3. I tried it also with a train-test split. However the same problem appears. The reason I shared the results using only the train set is so that I can compare the `$model` and `$score` results (the former being defined on the train set.)

**Comment by a-humble-learner** [comment]
Thanks @RaphaelS1 .
1) For prediction, `glmnet` uses lambda `1se` by default, which is `0.005779` here. I presume the `$model` output shows the mean C-index over CV folds for lambdas (or is it the C-index on the full training set?). In any case would they not also be comparing the predicted risk and the true times, otherwise how would these be defined?
3a) I shared the prediction$score on the train set for better comparison with $model... even if the output of $model is an average C-index over CV folds (which I haven't found a confirmation source yet) and not the C-index on the full training set, I would expect $score to be closer to $model? 
3b) On this tiny made-to-share dataset, train-test split may be enough to avoid $score = 1, but when I try train-test split on my real dataset $score on a holdout test set still returns 1, which is very surprising, would you not agree?

**Comment by RaphaelS1** [comment]
Hi @a-humble-learner if you resample the dataset with 3 folds you get the same cindex as in the glmnet model (see the first code block below). Secondly, there was indeed a bug that I have now fixed in https://github.com/mlr-org/mlr3proba/commit/0b9ea306a2966135eace7d6f2bb7d86974e31d50. You'll need to install latest version from main: `remotes::install_github(mlr-org/mlr3proba)`.

(P.S. This is why posting issues on GH is more efficient than email, much easier to find bugs and discuss individual issues. Thanks for spotting this one!)

``` r
library(mlr3proba)
library(mlr3learners)

setwd("~/Downloads")
df = read.csv("silly_data.csv")

df$cov1_ <- as.factor(df$cov1_)
mat <- as.data.frame(model.matrix(~., df, na.action = na.omit)[, -1])

task = TaskSurv$new(
  id = "testing",
  backend = as.data.frame(mat),
  time = "futime",
  event = "status"
)

learner = lrn("surv.cv_glmnet", alpha = 1, nfolds = 3, type.measure = "C")

learner$train(task)
learner$model
#> 
#> Call:  (if (cv) glmnet::cv.glmnet else glmnet::glmnet)(x = data, y = target,      type.measure = "C", nfolds = 3L, alpha = 1, family = "cox") 
#> 
#> Measure: C-index 
#> 
#>       Lambda Index Measure       SE Nonzero
#> min 0.004372    23  0.7219 0.005424       4
#> 1se 0.006342    19  0.7189 0.004719       4

resample(task, learner, rsmp("cv", folds = 3))$aggregate()
#> surv.cindex 
#>   0.7218067
```

<sup>Created on 2021-12-21 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>
``` r
library(mlr3learners)
library(mlr3proba)
setwd("~/Downloads")
df = read.csv("silly_data.csv")

df$cov1_ = as.factor(df$cov1_)
mat = as.data.frame(model.matrix(~., df, na.action = na.omit)[, -1])

task = as_task_surv(mat, time = "futime", event = "status")
learner = lrn("surv.cv_glmnet", alpha = 1, nfolds = 3, type.measure = "C")

learner$train(task)$predict(task)$score()
#> surv.cindex 
#>   0.7221327
```

<sup>Created on 2021-12-21 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by a-humble-learner** [comment]
Great, thanks for fixing it! I see it's a precision thing...
